core/src/main/java/org/infinispan/factories/DataContainerFactory.java

```diff
@@ -58,8 +58,8 @@
                if (clusteringConfiguration.cacheMode().needsStateTransfer()) {
                   int segments = clusteringConfiguration.hash().numSegments();
                   Supplier mapSupplier = () -> {
-                     OffHeapEntryFactory entryFactory = componentRegistry.getComponent(OffHeapEntryFactory.class);
-                     OffHeapMemoryAllocator memoryAllocator = componentRegistry.getComponent(OffHeapMemoryAllocator.class);
+                     OffHeapEntryFactory entryFactory = componentRegistry.getOrCreateComponent(OffHeapEntryFactory.class);
+                     OffHeapMemoryAllocator memoryAllocator = componentRegistry.getOrCreateComponent(OffHeapMemoryAllocator.class);
```

Comment:
I must say I don't like `getOrCreateComponent` here, IMO the logic in the supplier can be implemented in `DefaultSegmentedDataContainer` directly.

Reply:
Hrmm let me look again.

Reply:
I am not sure how you mean here. The `DefaultSegmentedDataContainer` is used for both off heap and on heap containers. This saves a ton of code duplication and fixed issues with expiration for off heap.

Can you share some code for how you recommend doing this? I defer to your registry wizardry :)

Reply:
No wizardry, just check the configuration in `DefaultSegmentedDataContainer` and create the appropriate maps...

Reply:
Don't forget that `DefaultSegmentedDataContainer` is used for the bounded off heap, bounded segmented off heap and bounded segmented on heap variants as well.

The class `DefaultSegmentedDataContainer` currently is a very simple lightweight reusable class that deals with just segmented implementation parts, I am not a fan of complicating it with all the config logic as well.

I would say if you want we can create a separate JIRA to look further into this.

Reply:
Either `DefaultSegmentedDataContainer` is very simple, or using it saved a ton of code duplication, you can't have both :)

OTOH I strongly believe component factories should be very simple and that `ComponentRegistry.get[OrCreate]Component()` should be used sparingly. With ISPN-9127 I'm making the components start only after their dependencies, and calling `get[OrCreate]Component()` hides the dependency.

Reply:
Okay :D

I live on the other side of the fence that the Components should be simple and config lives separately, properly separating concerns. I agree that Components should be Injected when possible though, unfortunately the class has different required components based on the config, heh.

Reply:
I don't want to over-complicate the components either, but I'd argue that having to call a supplier that can do anything, like instantiate another component, makes the component more complicated than injecting that component directly.